### DISCUSSION TITLE
How to update large amount of records?

### DISCUSSION THREAD

**Discussion by HEAVYPOLY** [pocketbase/pocketbase#3893] [Q&A] [answered]
I want to go through my users collection (1600 users) and replace the autogenerated usernames with the user's email before '@' like this:
I'm trying to run this while authenticated as admin but getting errors.
It works if I use getList(1,5)  but fails with full list (1600).

Does anyone know how to do this?

```js
async function setDefaultUsernames () {
  users = users.items
  let users = await pb.collection('users').getFullList()
  for (const user of users) {
    // console.log(user.email, user.username)
    let username = user.email.split('@')[0].replace('+', '.')
    try {
      console.timeLog("SETTING USERNAME SUCCESS")
      pb.collection('users').update(user.id, { username })
    } catch {
      console.log("ERROR SETTING USERNAME", username)
    }
  }
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
If you are "getting errors" please paste the exact errors here, otherwise I don't know how to help you debug it.
In any case I doubt that whatever error you are getting is from the update request because you are not awaiting the Promises and the `try/catch` will not have effect. Here is an example for sending concurrent requests and awaiting them all with `Promise.all` or `Promise.allSettled`:

```js
async function setDefaultUsernames () {
  const promises = [];

  const users = await pb.collection('users').getFullList()
  for (const user of users) {
    // console.log(user.email, user.username)
    let username = user.email.split('@')[0].replace('+', '.')

    promises.push(pb.collection('users').update(user.id, { username }))
  }

  return Promise.allSettled(promises)
}

setDefaultUsernames().then((results) =>
  results.forEach((result) => console.log(result.status)),
);
```

---

But your use case seems more suitable for a migration (Go or JS). Here is an example with the JS migrations:

1. Run `./pocketbase migrate create "set_default_usernames"`

2. Open the created file and use something like (_I haven't tested it_):
 ```js
  // pb_migrations/1701933182_set_default_usernames.js
  migrate((db) => {
    const dao = new Dao(db);

    // note: if you have a lot of users (eg. 100k+) you may want to get them on batches
    const users = arrayOf(new Record);
    dao.recordQuery("users").all(users);

    for (let user of users) {
      let email = user.get("email")
      user.set("username", email.split('@')[0].replace('+', '.'))
      dao.saveRecord(user)
    }
  }, null)
  ```
3. Start the application with the `serve` command as usual

**Comment by HEAVYPOLY** [comment]
Thanks for the quick answer.

With this setDefaultUsernames() I get:
```1618 rejected```


If I use getList(1,30):
```
11
fulfilled
main.js:27
2
rejected
main.js:27
fulfilled
main.js:27
2
rejected
main.js:27
3
fulfilled
main.js:27
rejected
main.js:27
3
fulfilled
main.js:27
2
rejected
main.js:27
2
fulfilled
main.js:27
rejected
main.js:27
2
fulfilled
```

This is the error I was getting before 
```
Uncaught Error ClientResponseError 0: Something went wrong while processing your request.
    at ClientResponseError (/Users/kakapoopie/webbypaint-admin/src/ClientResponseError.ts:13:9)
    at <anonymous> (/Users/kakapoopie/webbypaint-admin/src/Client.ts:310:23)
    --- await ---
    at <anonymous> (/Users/kakapoopie/webbypaint-admin/main.js:7:1)
ClientResponseError.ts:13
```

I'll try migration tomorrow.

**Reply by ganigeorgiev** [reply]
Are you sure that the result usernames are unique?

You can also start the executable with `--debug` and/or check the failed request logs in the _Admin UI > Logs_ for more details.

**Reply by ganigeorgiev** [reply]
The first output also seems to imply that your function is being executed multiple times.
Double check that you are not calling repeatedly the same method as in this case it will fail either because of the [autocancellation](https://github.com/pocketbase/js-sdk#auto-cancellation) or because of the unique usernames constraint.

In any case, if you are still not able to resolve it, please provide a more detailed code sample and elaborate a little more how and when your code is being executed.

**Reply by benallfree** [reply]
Since OP is doing this on https://pockethost.io it’s important also to remember that API requests are rate limited (50 per 10s per IP). 

I agree this is better done as a hook or migration. Uniqueness should be considered carefully, as well as what happens when users are created.

**Comment by HEAVYPOLY** [comment]
Migration was successful - thanks @benallfree and @ganigeorgiev for the help.

Now I am trying to figure out this hook so I can generate username from email for new users who use google oauth as well as email/pass signup:
```js
onModelBeforeCreate((e) => {
  console.log("SIGNUP HOOK!", e.model.email)
  e.model.username = $app.dao().suggestUniqueAuthRecordUsername('users', e.model.email.split(/@/)[0])
}, "users")
```

e.model.email is logged as reflect.methodValueCall() { [native code] } so the split fails.  How do I retrieve the email inside onModelBeforeCreate?

**Reply by ganigeorgiev** [reply]
> How do I retrieve the email inside onModelBeforeCreate?

`e.model` in this case is a `Record` pointer, aka. it is not a plain JS object, and you need to use the [get/set Record model methods](https://pocketbase.io/docs/js-records/#getset-record-fields):
```js
e.model.set("username", $app.dao().suggestUniqueAuthRecordUsername("users", e.model.get("email")....))
```

**Comment by HEAVYPOLY** [comment]
It works!  Thank you again

**Comment by HEAVYPOLY** [comment]
I'm also interested in populating avatar and name from existing oAuth users via migration and on new oAuth signup.  Is this possible or is the oAuth data only available when the user is created?
```js

migrate((db) => {
  const dao = new Dao(db);
  const users = arrayOf(new Record);
  dao.recordQuery("users").all(users);
  //GOOGLE OAUTH
  for (let user of users) {
    let email = user.get("email")
    let avatarFromGoogle = user.get("avatarUrl")
    let nameFromGoogle = user.get("Name") 
    if (avatarFromGoogle) user.set("avatar", avatarFromGoogle)
    if (nameFromGoogle) user.set("name", nameFromGoogle)
    if (avatarFromGoogle || nameFromGoogle) dao.saveRecord(user)
  }
}, null)
```

**Reply by ganigeorgiev** [reply]
It is not possible to access the OAuth2 user data from a migration because the only thing we store from the OAuth2 auth response is the OAuth2 user id that we link to the PocketBase user.

If you want to save the OAuth2 avatar as PocketBase user image file, you can attach to the `onRecordAfterAuthWithOAuth2Request` hook, download the avatar url and resave the related record model. I haven't tested it, but here is one such example:

```js
onRecordAfterAuthWithOAuth2Request((e) => {
    if (
        // no oauth avatar
        e.oAuth2User.avatarUrl == "" ||
        // the record already has assigned avatar
        e.record.get("avatar") != ""
    ) {
        return;
    }

    // download the oauth2 avatar
    const res = $http.send({
        url: e.oAuth2User.avatarUrl,
    })
    if (res.statusCode != 200) {
        return; // failed to download the avatar image
    }

    // get the avatar name from the last part of the url path
    const avatarFilename = e.oAuth2User.avatarUrl.split("/").pop();

    // init a record upsert form to load the downloaded avatar file
    // so that you can validate it and safely persist it into the database
    const form = new RecordUpsertForm($app, e.record)
    form.addFiles("avatar", $filesystem.fileFromBytes(res.raw, avatarFilename))
    form.submit();
}, "users")
```

**Reply by HEAVYPOLY** [reply]
Thanks, will try this.
